nrpy/infrastructures/BHaH/general_relativity/BSSN/psi4_decomposition.py

```diff
@@ -3,161 +3,542 @@
+        __name__,
+        BHd_str,
+    )
+
+
+def register_CFunction_psi4_spinweightm2_decomposition(CoordSystem: str) -> None:
+    """
+    Register C function for decomposing psi4 into spin-weighted spherical harmonics via 2D interpolation on uniform source grids at each φ slice.
+
+    :param CoordSystem: Specifies the coordinate system for psi4 decomposition.
+    :raises ValueError: If psi4 decomposition is not supported for the coordinate system.
+    """
+    if "Spherical" in CoordSystem:
+        radial_like_index = 0
+        theta_like_index = 1
+        phi_index = 2
+    elif "Cylindrical" in CoordSystem:
+        radial_like_index = 0
+        theta_like_index = 2
+        phi_index = 1
+    elif "SymTP" in CoordSystem:
+        radial_like_index = 0
+        theta_like_index = 1
+        phi_index = 2
+    else:
+        raise ValueError(f"CoordSystem = {CoordSystem} not supported.")
+
+    setup_prefunc = rf"""
+// IDX2GENERAL: Map 2D indices (i,j) to a 1D index using stride Ni
+#define IDX2GENERAL(i, j, Ni) ((i) + (Ni) * (j))
+// REVERSE_IDX2GENERAL: Recover (i,j) from a 1D index and stride Ni
+#define REVERSE_IDX2GENERAL(index, Ni, i, j) \
+{{                                            \
+    j = (index) / (Ni);                      \
+    i = (index) % (Ni);                      \
+}}
+
+/*
+This prefunction sets up the diagnotic struct:
+
+typedef struct __diagnostic_struct__ {{
+  int *restrict N_shell_pts_grid;            // of shape int [NUM_OF_R_EXTS]
+  REAL **restrict xx_radial_like_shell_grid; // of shape [NUM_OF_R_EXTS][N_shell_pts_grid]
+  REAL **restrict xx_theta_like_shell_grid;  // of shape [NUM_OF_R_EXTS][N_shell_pts_grid]
+  int *restrict N_theta_shell_grid;          // of shape int [NUM_OF_R_EXTS]
+  REAL **restrict theta_shell_grid;          // of shape [NUM_OF_R_EXTS][N_theta_shell_grid]
+  REAL dtheta;
+}} diagnostic_struct;
+
+Set up thin shells at R_ext radii for integration of psi4 * spin weight 2 spherical harmonics.
+
+Assumes that grid might be a rectangular partition of the whole grid.
+
+For each shell the # of shell points which lie on the grid "N_shell_pts_grid" is found.
+
+The destination points for 2d interpolation  "xx_radial_like_shell_grid" and ""xx_theta_like_shell_grid" are computed.
+*/
+static void psi4_diagnostics_set_up(const commondata_struct *restrict commondata, const params_struct *restrict params, REAL *restrict xx[3],
+                                                     diagnostic_struct *restrict diagnosticstruct) {{
+
+  const REAL *restrict list_of_R_exts = commondata->list_of_R_exts;
+
+  // Thin shells with N_theta and N_phi number of points are constructed at each extraction radius R_ext
+  const int N_theta = commondata->num_theta_points_on_shell_for_psi4_interp;
+
+  // Phi values of shell need to be exactly the same as phi values of grid, all coordinate systems supported have a phi coordinate
+  const int N_phi = params->Nxx{phi_index};
```

Comment:
For easier maintainability, I would suggest doing something like setting a static constant integer like:
`static const int phi_dirn = {phi_dirn};`
just prior to this prefunc.

Then I would define
```
const int Nxx[3] = { params->Nxx0, params->Nxx1, params->Nxx2 };
const int N_phi = Nxx[phi_dirn];
```

The result would enable you to remove the f-string (f""") at the top of this string definition, and make the entire C code much more maintainable.